### PR TITLE
do gates in one go

### PR DESCRIPTION
**Pull request by szagoruyko**
a trick from @skaae https://gist.github.com/skaae/ea4320e17379d408e693 completes short training almost 2 times faster, in 34 minutes.

### CODE REVIEW

**Comment by ghost** [comment]
Hey @szagoruyko Andrej Karpathy just posted a fast looking LSTM implementation [here](https://gist.github.com/karpathy/7bae8033dcf5ca2630ba)

Do you want to try it? 

I'll have a ago as well, [on my fork](https://github.com/AjayTalati/lstm/blob/master/main.lua), and we can compare results? 

I have a slow GPU, (2.0 compute capability, and hence cannot get `fbcunn` to install), so my comparison will not be very good ?

I just did the experiment and both versions take about 46 mins. For some reason though Andrej Karpathy's code got a lower perplexity in both validation and test sets? Did you find this?

**Comment by szagoruyko** [comment]
So on Titan Black @skaae took 34 min https://gist.github.com/szagoruyko/b4b153f52e1a8a95bd16
@karpathy's took 37 min https://gist.github.com/szagoruyko/9f3c0e41732ab485c602
I like skaae implementation more

**Comment by ghost** [comment]
@szagoruyko THANKS MAN !!!!

I'm going to the shops now to buy a Titan Black :+1: 

PS -  was there any significant difference in perplexity scores - I know it does not make sense to compare over just one sample path - I'm just curious?

**Comment by soumith** [comment]
@AjayTalati around 15 mins on Titan-X

**Comment by ghost** [comment]
@soumith Thanks a lot dude :+1:  

Did you try Andrej Karpathy's [LSTM code](https://github.com/AjayTalati/lstm/blob/master/main.lua), as well?

OK so I should get a Titan- X :) man that's FAST !!!!  

Maybe trying 10 or 20 sample runs each with different seeds and comparing the averaged perplexity scores for both implementations - is an experiment worth doing? Should be straightforward to do using,

`cutorch.manualSeedAll( sample_run_number )`

Edit - I tried the above line to set the seed but it doesn't seem to work :-1: - got exactly the same perplexities as before? This is perplexing ???

**Comment by skaae** [comment]
I'm not 100% sure that the slicing operation is correct. The "trick" is that you calculate all gates in a single matrix operation and then slice the results. Instead of having a weight matrix for each gate you have single matrix that is shape `num_units, 4*num_units`. 

In Numpy i have implemented this as:

``` Python
import numpy as np
num_units = 10
batch_size = 5
n_gates = 4

hid_prev = np.random.random((batch_size, num_units))
W_hid_to_hid = np.random.random((num_units, n_gates*num_units))

def slice_w(x, n):
       return x[:, n*num_units:(n+1)*num_units]


# gates not including non linearities    
gates = np.dot(hid_prev, W_hid_to_hid)
ingate = slice_w(gates, 0)
forgetgate = slice_w(gates, 1)
modulationgate = slice_w(gates, 2)
outgate = slice_w(gates, 3)
```

I torch I implemented the slicing operation as a reshape and a split operation:

``` LUA
batch_size = 4
num_units = 3
n_gates = 4
mlp=nn.Sequential()
mlp:add(nn.Reshape(n_gates,num_units))  -- reshape to (batch_size, n_gates, num_uints)
mlp:add(nn.SplitTable(2))               -- split on n_gates dimension


x = torch.Tensor(batch_size,n_gates*num_units)
for i = 1, batch_size do
   for j = 1, n_gates*num_units do
       x[i][j] = (i-1)*4+j
    end
 end
print(x)
pred=mlp:forward(x)

for i = 1, n_gates do
    print(pred[i])
end

```

You might be able to speed up the calculation further by precalculating 
`i2h = nn.Linear(opt.rnn_size, 4*opt.rnn_size)` outside of the for loop.

**Comment by wojzaremba** [comment]
I have merged it. I am a terrible git user, so I had to copy your code, and commit it by myself. I am not sure how to accept a pull request. Thank you szagoruyko a lot. Good job!

**Comment by skaae** [comment]
@wojzaremba: Dont you have a large green merge pull request at the bottom of your page?

**Comment by szagoruyko** [comment]
a small note to this, it takes 22 minutes on Titan Black with @tkocisky's oxnn, the code is here https://github.com/szagoruyko/lstm/blob/oxnn/main.lua
